### PR TITLE
pyomo.gdp: Fixing bugs in nested models in gdp.hull transformation

### PR DESCRIPTION
**Pull request by emma58**
## Summary/Motivation:

This fixes several bugs in hull having to do with nested GDPs:
1. Primarily, it corrects the prior implementation's wrong assumption that all nested indicator variables are local. This doesn't have to be true, and when it's not, these variables need to be disaggregated like anything else.
2. It also corrects a bug in variable disaggregation in nested GDPs where some layers of the hierarchy don't use variables that appear as decendents to them in GDP trees.
3. There was a silly typo bug having to do with disaggregation too.

Most notably, in order to accomplish the above, this switches the hull transformation from transforming from root to leaf in the GDP tree to the opposite. Note that this is the reason that the baselines changed and that the FME tests changed (and can be blamed for lots of other chaos as well).

## Changes proposed in this PR:
- Rewrites hull to transform from leaf to root in the GDP tree
- Corrects handling of nested structures in terms of disaggregated variables, particularly not assuming any indicator variables are local unless it is declared
- Adds/modifies a lot of tests for nested GDPs
- Partially moves the mappings between original and transformed components to the new `private_data` magic, though does not do this completely yet--I will save that for a later PR

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
5. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/gdp/plugins/hull.py [hunk 1]
```diff
@@ -648,54 +651,40 @@ def _declare_disaggregated_var_bounds(
         if ub:
             bigmConstraint.add(ub_idx, disaggregatedVar <= ub * var_free_indicator)
 
+        original_var_info = original_var.parent_block().private_data()
+        if 'disaggregated_var_map' not in original_var_info:
+            original_var_info['disaggregated_var_map'] = ComponentMap()
+        disaggregated_var_map = original_var_info['disaggregated_var_map']
```
Comment (jsiirola) [inline pyomo/gdp/plugins/hull.py:657]: We should add a `default=` or `initialize=` argument to `private_data()` that the method can use to initialize the private dict if it is not present.

File: pyomo/gdp/util.py [hunk 2]
```diff
@@ -169,7 +169,10 @@ def parent_disjunct(self, u):
         Arg:
             u : A node in the forest
         """
-        return self.parent(self.parent(u))
+        if isinstance(u, _DisjunctData) or u.ctype is Disjunct:
```
[the remark this answers is not in the document]
Comment (emma58) [reply pyomo/gdp/util.py:172]: Yeah, I think so.

File: pyomo/gdp/plugins/hull.py [hunk 3]
```diff
@@ -204,33 +199,41 @@ def __init__(self):
...
+                        self._collect_local_vars_from_block(b, user_defined_local_vars)
+                        seen_blocks.add(b)
+                # now look up in the tree
+                blk = t
+                while blk is not None:
+                    if blk not in seen_blocks:
+                        self._collect_local_vars_from_block(
+                            blk, user_defined_local_vars
+                        )
+                        seen_blocks.add(blk)
```
[the remark this answers is not in the document]
Comment (emma58) [reply pyomo/gdp/plugins/hull.py]: Yes.

File: pyomo/gdp/plugins/hull.py (hunk 1 above)
Comment (jsiirola) [reply pyomo/gdp/plugins/hull.py:650]: Scratch that - we implemented this as an explicit registration: #3153

**Review by jsiirola — APPROVED** [review]

**Pull request merged**